near-sdk-macros/src/core_impl/utils/mod.rs

```diff
@@ -73,3 +75,62 @@ pub(crate) fn extract_vec_type(ty: &Type) -> Option<&Type> {
         _ => None,
     }
 }
+
+fn _sanitize_self(typ: TokenStream2, replace_with: &TokenStream2) -> TokenStream2 {
+    let trees = typ.into_iter().map(|t| match t {
+        TokenTree::Ident(ident) if ident == "Self" => {
+            let replace_with = replace_with
+                .clone()
+                .into_iter()
+                .map(|mut t| {
+                    t.set_span(ident.span());
+                    t
+                })
+                .collect();
+            TokenTree::Group(Group::new(Delimiter::None, replace_with))
+        }
+        TokenTree::Group(group) => {
+            let stream = _sanitize_self(group.stream(), replace_with);
+            TokenTree::Group(Group::new(group.delimiter(), stream))
+        }
+        rest => rest,
+    });
+    trees.collect()
```

Comment:
Not a big deal at all, but I think you can avoid creating some `TokenTree`s here if you operate on `impl Iterator<Item = TokenStream2>` instead.

Reply:
`TokenTree` really is what helps us with this level of granularity, though.

Reply:
WDYM? Couldn't you do something like (untested):

```rust
    let trees = typ.into_iter().map(|t| match t {
        TokenTree::Ident(ident) if ident == "Self" => replace_with
            .clone()
            .into_iter()
            .map(|mut t| {
                t.set_span(ident.span());
                t
            })
            .collect::<TokenStream2>(),
        TokenTree::Group(group) => {
            let stream = _sanitize_self(group.stream(), replace_with);
            TokenTree::Group(Group::new(group.delimiter(), stream)).into()
        }
        rest => rest.into(),
    });
    trees.collect()
```

Reply:
Yeah, tried that approach in [`6df05aa` (#1001)](https://github.com/near/near-sdk-rs/pull/1001/commits/6df05aad394007e506a1664c7a4b561e801d975b)? Hard to get over the `TokenTree` from the group part. Another issue might also just be the unnecessary allocation that `TokenStream` does behind the scenes, whereas operating on `TokenTree`'s esp in an iterator context can be more efficient.

Reply:
> Another issue might also just be the unnecessary allocation that TokenStream does behind the scenes

Hmm, yeah I peeked the `FromIterator<TokenTree>` implementation and it seems like they do a bunch of optimizations there. Probably more efficient indeed, but your call in the end.

Reply:
Well, the change is small, and I don't foresee any sufficiently complex type signatures that cause OOMs 😆 
So we'll leave it as is.